### ISSUE TITLE
Serialization is incorrect with sequences

### ISSUE THREAD

**Issue by sumanth-nirmal** [eclipse-cyclonedds/cyclonedds-cxx#121]
## Required information

**Operating system:**
Ubuntu 18.04 LTS

**Cyclone version:**
- cyclonedds - master, [7bc43](https://github.com/eclipse-cyclonedds/cyclonedds/commit/7bc432683690ba860a72c662395c07835ce0b3a4)
- cyclonedds-cxx - master, ca478d5d1e3b1cfe6b5791248feb7d79d4af3ce0 rebased onto branch [reicheratwork/serialization-fixes](https://github.com/reicheratwork/cyclonedds-cxx/tree/serialization-fixes)

**Observed result or behaviour:**

The serialization/de-serialization doesn't seem to happen correctly with sequences

**Detailed reproducer**

When using the following nested IDL files

- [BasicTypes_.idl.txt](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/files/6754716/BasicTypes_.idl.txt)
- [Constants_.idl.txt](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/files/6754720/Constants_.idl.txt)
- [Defaults_.idl.txt](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/files/6754731/Defaults_.idl.txt)
- [UnboundedSequences_.idl.txt](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/files/6754735/UnboundedSequences_.idl.txt)
 (I have used `.txt` extension, since git doesn't allow me to upload `.idl` files)

in the code as below:
```cpp
std::shared_ptr<test_msgs::msg::dds_::basic_types>
get_msg_basic_types()
{
  auto msg = std::make_shared<test_msgs::msg::dds_::basic_types>();
  {
    msg->bool_value_(false);
    msg->byte_value_(0);
    msg->char_value_(0);
    msg->float32_value_(0.0f);
    msg->float64_value_(0);
    msg->int8_value_(0);
    msg->uint8_value_(0);
    msg->int16_value_(0);
    msg->uint16_value_(0);
    msg->int32_value_(0);
    msg->uint32_value_(0);
    msg->int64_value_(0);
    msg->uint64_value_(0);
  }
  return msg;
}

std::shared_ptr<test_msgs::msg::dds_::unbounded_seq>
get_msg_unbounded_sequences()
{
  auto basic_types_msgs = get_msg_basic_types();
  auto msg = std::make_shared<test_msgs::msg::dds_::unbounded_seq>();
  {
    msg->bool_values_({true});
    msg->byte_values_({0xff});
    msg->char_values_({255});
    msg->float32_values_({1.125f});
    msg->float64_values_({1.125});
    msg->int8_values_({(std::numeric_limits<int8_t>::max)()});
    msg->uint8_values_({(std::numeric_limits<uint8_t>::max)()});
    msg->int16_values_({(std::numeric_limits<int16_t>::max)()});
    msg->uint16_values_({(std::numeric_limits<uint16_t>::max)()});
    msg->int32_values_({(std::numeric_limits<int32_t>::max)()});
    msg->uint32_values_({(std::numeric_limits<uint32_t>::max)()});
    msg->int64_values_({(std::numeric_limits<int64_t>::max)()});
    msg->uint64_values_({(std::numeric_limits<uint64_t>::max)()});
    msg->string_values_({{"max value"}});
    msg->basic_types_values_({*basic_types_msgs});
    msg->alignment_check_(1);
  }
  return msg;
}

TEST(tooch, serialize_deserialize)
{
  using DDSType = test_msgs::msg::dds_::unbounded_seq;

  // data
  auto msg = get_msg_unbounded_sequences();

  org::eclipse::cyclonedds::core::cdr::basic_cdr_stream cdr_streamer;
  uint8_t buffer[2048];
  // fill the buffer to some non-zero to see the effect
  std::memset(buffer, 0xaa, sizeof(buffer));
  {
    //serialize
    cdr_streamer.set_buffer(buffer);
    org::eclipse::cyclonedds::core::cdr::write(cdr_streamer, *msg);
    EXPECT_FALSE(cdr_streamer.abort_status());
  }

  DDSType deserialized_msg{};
  EXPECT_NE(*msg, deserialized_msg);
  {
    //deserialize
    cdr_streamer.set_buffer(buffer);
    org::eclipse::cyclonedds::core::cdr::read(cdr_streamer, deserialized_msg);
    EXPECT_FALSE(cdr_streamer.abort_status());
  }
  EXPECT_EQ(*msg, deserialized_msg);
}
```
In the code above, I try to serialize the type, then deserialize it back and compare it with the original type, which should ideally match with the original type, but it fails with the below error:

```sh
Expected equality of these values:
  *msg
    Which is: 784-byte object <A0-54 68-94 A9-55 00-00 00-00 00-00 00-00 00-00 A0-54 68-94 A9-55 00-00 01-00 00-00 00-00 00-00 A8-54 68-94 A9-55 00-00 C0-54 68-94 A9-55 00-00 C1-54 68-94 A9-55 00-00 C1-54 68-94 A9-55 00-00 ... 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 01-00 00-00 00-00 00-00>
  deserialized_msg
    Which is: 784-byte object <90-D1 61-94 A9-55 00-00 00-00 00-00 00-00 00-00 90-D1 61-94 A9-55 00-00 01-00 00-00 00-00 00-00 98-D1 61-94 A9-55 00-00 70-D1 61-94 A9-55 00-00 71-D1 61-94 A9-55 00-00 71-D1 61-94 A9-55 00-00 ... 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 01-00 00-00 00-00 00-00>
```

**Expected result or behaviour:**

The serialization/deserialization should work correctly with all types of sequences

**Comment by sumanth-nirmal** [comment]
@reicheratwork I think I found the issue, it is not about the sequence, but how the check here https://github.com/eclipse-cyclonedds/cyclonedds-cxx/blob/master/src/idlcxx/src/streamers.c#L280 is done to verify if the type is `bool`

Because of this check, `uint64` and `int64` are treated as bools, the generated code for the `int64`, `uint64` looks like below

```hpp
module msg
{
struct log {
sequence<boolean> bool_values_;
sequence<long long> int64_values_;
squence<unsigned long long> uint64_values_;
};
};
```
```hpp
...
template<typename T>
void write(T& streamer, const ::msg::log& instance)
{
  {
  uint32_t se_1 = uint32_t(instance.bool_values_().size());
  write(streamer, se_1);
  for (uint32_t i_1 = 0; i_1 < se_1; i_1++) {
  {
    const bool b = instance.bool_values_()[i_1];
    write(streamer, b);
  }
  } //i_1
  }
  {
  uint32_t se_1 = uint32_t(instance.int64_values_().size());
  write(streamer, se_1);
  for (uint32_t i_1 = 0; i_1 < se_1; i_1++) {
  {
    const bool b = instance.int64_values_()[i_1];
    write(streamer, b);
  }
  } //i_1
  }
  {
  uint32_t se_1 = uint32_t(instance.uint64_values_().size());
  write(streamer, se_1);
  for (uint32_t i_1 = 0; i_1 < se_1; i_1++) {
  {
    const bool b = instance.uint64_values_()[i_1];
    write(streamer, b);
  }
  } //i_1
  }
  (void)streamer;
  (void)instance;
}
...
```
This is incorrect and the correct `write()` should be as below

```hpp
template<typename T>
void write(T& streamer, const ::msg::log& instance)
{
  {
  uint32_t se_1 = uint32_t(instance.bool_values_().size());
  write(streamer, se_1);
  for (uint32_t i_1 = 0; i_1 < se_1; i_1++) {
  {
    const bool b = instance.bool_values_()[i_1];
    write(streamer, b);
  }
  } //i_1
  }
  {
  uint32_t se_1 = uint32_t(instance.int64_values_().size());
  write(streamer, se_1);
  for (uint32_t i_1 = 0; i_1 < se_1; i_1++) {
  write(streamer, instance.int64_values_()[i_1]);
  } //i_1
  }
  {
  uint32_t se_1 = uint32_t(instance.uint64_values_().size());
  write(streamer, se_1);
  for (uint32_t i_1 = 0; i_1 < se_1; i_1++) {
  write(streamer, instance.uint64_values_()[i_1]);
  } //i_1
  }
  (void)streamer;
  (void)instance;
}
```

This messes up the serialization/de-serialization when `int64`/`uint64` are involved in te IDL types, we should probably fix the check to something like below

```diff
-  if ((idl_mask(type_spec) & IDL_BOOL) == IDL_BOOL &&
+  if ((idl_mask(type_spec)) == IDL_BOOL &&
```

Does it make sense?

**Comment by k0ekk0ek** [comment]
That'll work just fine, but it's better to use `idl_type(type_spec)`. That'll return a valid enum value or `IDL_NULL`.

**Comment by sumanth-nirmal** [comment]
@k0ekk0ek Thanks, I fixed it according to your suggestion here #124
